# Property Price Prediction

## Load the libraries

In [53]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# skip warnings
import warnings
warnings.filterwarnings('ignore')
warnings.filterwarnings(action = 'ignore', category = DeprecationWarning)
warnings.filterwarnings(action = 'ignore', category = FutureWarning)

pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',None)  
pd.set_option('display.expand_frame_repr',False)
pd.set_option('max_colwidth',-1)

## Read data

In [54]:
data = pd.read_csv('Property_Price.csv')

In [55]:
data.head()

Id  Building_Class Zoning_Class  Lot_Extent  Lot_Size Road_Type Lane_Type Property_Shape Land_Outline Utility_Type Lot_Configuration Property_Slope Neighborhood Condition1 Condition2 House_Type House_Design  Overall_Material  House_Condition  Construction_Year  Remodel_Year Roof_Design Roof_Quality Exterior1st Exterior2nd Brick_Veneer_Type  Brick_Veneer_Area Exterior_Material Exterior_Condition Foundation_Type Basement_Height Basement_Condition Exposure_Level BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  Total_Basement_Area Heating_Type Heating_Quality Air_Conditioning Electrical_System  First_Floor_Area  Second_Floor_Area  LowQualFinSF  Grade_Living_Area  Underground_Full_Bathroom  Underground_Half_Bathroom  Full_Bathroom_Above_Grade  Half_Bathroom_Above_Grade  Bedroom_Above_Grade  Kitchen_Above_Grade Kitchen_Quality  Rooms_Above_Grade Functional_Rate  Fireplaces Fireplace_Quality  Garage  Garage_Built_Year Garage_Finish_Year  Garage_Size  Garage_Area Garage_Quality Garage_Condition Pavedd_Drive  W_Deck_Area  Open_Lobby_Area  Enclosed_Lobby_Area  Three_Season_Lobby_Area  Screen_Lobby_Area  Pool_Area Pool_Quality Fence_Quality Miscellaneous_Feature  Miscellaneous_Value  Month_Sold  Year_Sold Sale_Type Sale_Condition  Sale_Price
0  1   60              RLD          65.0        8450      Paved     NaN       Reg            Lvl          AllPub       I                 GS             CollgCr      Norm       Norm       1Fam       2Story       7                 5                2003               2003          Gable       SS           VinylSd     VinylSd     BrkFace           196.0              Gd                TA                 PC              Gd              TA                 No             GLQ          706         Unf          0           150        856                  GasA         Ex              Y                SBrkr             856               854                0             1710               1                          0                          2                          1                          3                    1                    Gd              8                  TF              0           NaN               Attchd  2003.0             RFn                2            1085.793744  TA             TA               Y            163.788080   69.596115        20.337934            0                        0                  0          NaN          NaN           NaN                   0                    2           2008       WD        Normal         208500    
1  2   20              RLD          80.0        9600      Paved     NaN       Reg            Lvl          AllPub       FR2P              GS             Veenker      Feedr      Norm       1Fam       1Story       6                 8                1976               1976          Gable       SS           MetalSd     MetalSd     None              0.0                TA                TA                 CB              Gd              TA                 Gd             ALQ          978         Unf          0           284        1262                 GasA         Ex              Y                SBrkr             1262              0                  0             1262               0                          1                          2                          0                          3                    1                    TA              6                  TF              1           TA                Attchd  1976.0             RFn                2            196.316304   TA             TA               Y            198.900074   74.716033        15.039392            0                        0                  0          NaN          NaN           NaN                   0                    5           2007       WD        Normal         181500    
2  3   60              RLD          68.0        11250     Paved     NaN       IR1            Lvl          AllPub       I                 GS             CollgCr      Norm       Norm       1Fam       2Sto

In [56]:
data.shape

(1459, 81)

In [57]:
data.Id.nunique()

1459

As all the values in this column are unique, we will **set Id as row index.**

In [58]:
data.set_index('Id', inplace = True)

In [59]:
data.describe().T

count           mean           std           min            25%            50%            75%            max
Building_Class             1459.0  56.922550      42.304033     20.000000     20.000000      50.000000      70.000000      190.000000   
Lot_Extent                 1200.0  70.045833      24.294456     21.000000     59.000000      69.000000      80.000000      313.000000   
Lot_Size                   1459.0  10517.225497   9984.675721   1300.000000   7549.000000    9477.000000    11603.000000   215245.000000
Overall_Material           1459.0  6.100069       1.383171      1.000000      5.000000       6.000000       7.000000       10.000000    
House_Condition            1459.0  5.575051       1.113125      1.000000      5.000000       5.000000       6.000000       9.000000     
Construction_Year          1459.0  1971.272104    30.212814     1872.000000   1954.000000    1973.000000    2000.000000    2010.000000  
Remodel_Year               1459.0  1984.879369    20.645927     1950.000000   1967.000000    1994.000000    2004.000000    2010.000000  
Brick_Veneer_Area          1451.0  103.756720     181.108150    0.000000      0.000000       0.000000       166.000000     1600.000000  
BsmtFinSF1                 1459.0  443.374914     456.142186    0.000000      0.000000       383.000000     712.000000     5644.000000  
BsmtFinSF2                 1459.0  46.382454      161.248500    0.000000      0.000000       0.000000       0.000000       1474.000000  
BsmtUnfSF                  1459.0  567.535984     441.874057    0.000000      223.500000     479.000000     808.000000     2336.000000  
Total_Basement_Area        1459.0  1057.293352    438.824912    0.000000      795.500000     991.000000     1298.500000    6110.000000  
First_Floor_Area           1459.0  1162.562714    386.712553    334.000000    882.000000     1086.000000    1391.500000    4692.000000  
Second_Floor_Area          1459.0  347.230295     436.583480    0.000000      0.000000       0.000000       728.000000     2065.000000  
LowQualFinSF               1459.0  5.848526       48.639512     0.000000      0.000000       0.000000       0.000000       572.000000   
Grade_Living_Area          1459.0  1515.641535    525.616607    334.000000    1129.000000    1464.000000    1777.500000    5642.000000  
Underground_Full_Bathroom  1459.0  0.424949       0.518870      0.000000      0.000000       0.000000       1.000000       3.000000     
Underground_Half_Bathroom  1459.0  0.057574       0.238830      0.000000      0.000000       0.000000       0.000000       2.000000     
Full_Bathroom_Above_Grade  1459.0  1.565456       0.550906      0.000000      1.000000       2.000000       2.000000       3.000000     
Half_Bathroom_Above_Grade  1459.0  0.382454       0.502798      0.000000      0.000000       0.000000       1.000000       2.000000     
Bedroom_Above_Grade        1459.0  2.866347       0.816050      0.000000      2.000000       3.000000       3.000000       8.000000     
Kitchen_Above_Grade        1459.0  1.046607       0.220410      0.000000      1.000000       1.000000       1.000000       3.000000     
Rooms_Above_Grade          1459.0  6.518163       1.625894      2.000000      5.000000       6.000000       7.000000       14.000000    
Fireplaces                 1459.0  0.613434       0.644687      0.000000      0.000000       1.000000       1.000000       3.000000     
Garage_Built_Year          1378.0  1978.515965    24.696004     1900.000000   1961.000000    1980.000000    2002.000000    2010.000000  
Garage_Size                1459.0  1.767649       0.747301      0.000000      1.000000       2.000000       2.000000       4.000000     
Garage_Area                1459.0  470.934787     210.668787   -129.369350    329.290489     471.028116     607.184977     1147.488093  
W_Deck_Area                1459.0  93.015667      124.796682   -338.112031    9.656026       92.803628      180.335986     572.298709   
Open_Lobby_Area            1459.0  47.811902      67.46

**Observation:**
    
    The wide difference between min and max values tell us that we have few outliers in our data.
    
    we can also drop all those columns where 75% of our observations is zero.
    

## check for missing values

In [60]:
data.isnull().sum()

Building_Class               0   
Zoning_Class                 0   
Lot_Extent                   259 
Lot_Size                     0   
Road_Type                    0   
Lane_Type                    1368
Property_Shape               0   
Land_Outline                 0   
Utility_Type                 0   
Lot_Configuration            0   
Property_Slope               0   
Neighborhood                 0   
Condition1                   0   
Condition2                   0   
House_Type                   0   
House_Design                 0   
Overall_Material             0   
House_Condition              0   
Construction_Year            0   
Remodel_Year                 0   
Roof_Design                  0   
Roof_Quality                 0   
Exterior1st                  0   
Exterior2nd                  0   
Brick_Veneer_Type            8   
Brick_Veneer_Area            8   
Exterior_Material            0   
Exterior_Condition           0   
Foundation_Type              0   
Basement_Heigh

## Display only missing columns in percentage

In [61]:
missing = data.isnull().sum() *100 / len(data)
missing = missing [missing > 0]
missing.sort_values(inplace = True)
missing

Electrical_System        0.068540 
Brick_Veneer_Type        0.548321 
Brick_Veneer_Area        0.548321 
Basement_Height          2.535984 
Basement_Condition       2.535984 
BsmtFinType1             2.535984 
Exposure_Level           2.604524 
BsmtFinType2             2.604524 
Garage_Condition         5.551748 
Garage_Quality           5.551748 
Garage_Finish_Year       5.551748 
Garage                   5.551748 
Garage_Built_Year        5.551748 
Lot_Extent               17.751885
Fireplace_Quality        47.224126
Fence_Quality            80.740233
Lane_Type                93.762851
Miscellaneous_Feature    96.298835
Pool_Quality             99.520219
dtype: float64

Dropping of those columns which has more than 45% of missing data

In [62]:
data.columns[data.isnull().sum()*100/data.shape[0] > 45]

Index(['Lane_Type', 'Fireplace_Quality', 'Pool_Quality', 'Fence_Quality',
       'Miscellaneous_Feature'],
      dtype='object')

In [63]:
data = data.drop(['Lane_Type', 'Fireplace_Quality', 'Pool_Quality', 'Fence_Quality',
                   'Miscellaneous_Feature'], axis = 1)

In [64]:
data.shape

(1459, 75)

In [65]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1459 entries, 1 to 1459
Data columns (total 75 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Building_Class             1459 non-null   int64  
 1   Zoning_Class               1459 non-null   object 
 2   Lot_Extent                 1200 non-null   float64
 3   Lot_Size                   1459 non-null   int64  
 4   Road_Type                  1459 non-null   object 
 5   Property_Shape             1459 non-null   object 
 6   Land_Outline               1459 non-null   object 
 7   Utility_Type               1459 non-null   object 
 8   Lot_Configuration          1459 non-null   object 
 9   Property_Slope             1459 non-null   object 
 10  Neighborhood               1459 non-null   object 
 11  Condition1                 1459 non-null   object 
 12  Condition2                 1459 non-null   object 
 13  House_Type                 1459 non-null   objec

In [66]:
cat_null_cols = ['Electrical_System', 'Brick_Veneer_Type', 'Basement_Height', 'Basement_Condition',
                 'BsmtFinType1', 'Exposure_Level', 'BsmtFinType2', 'Garage_Condition', 'Garage_Quality',
                 'Garage_Finish_Year', 'Garage']

In [67]:
data.Electrical_System.value_counts()

SBrkr    1333
FuseA    94  
FuseF    27  
FuseP    3   
Mix      1   
Name: Electrical_System, dtype: int64

In [68]:
data.Brick_Veneer_Type.value_counts()

None       863
BrkFace    445
Stone      128
BrkCmn     15 
Name: Brick_Veneer_Type, dtype: int64

In [69]:
data.Basement_Height.value_counts()

TA    648
Gd    618
Ex    121
Fa    35 
Name: Basement_Height, dtype: int64

In [70]:
data.Basement_Condition.value_counts()

TA    1310
Gd    65  
Fa    45  
Po    2   
Name: Basement_Condition, dtype: int64

In [71]:
data.BsmtFinType1.value_counts()

Unf    430
GLQ    418
ALQ    220
BLQ    147
Rec    133
LwQ    74 
Name: BsmtFinType1, dtype: int64

In [72]:
data.Exposure_Level.value_counts()

No    952
Av    221
Gd    134
Mn    114
Name: Exposure_Level, dtype: int64

In [73]:
data.BsmtFinType2.value_counts()

Unf    1256
Rec    54  
LwQ    45  
BLQ    33  
ALQ    19  
GLQ    14  
Name: BsmtFinType2, dtype: int64

In [74]:
data.Garage_Condition.value_counts()

TA    1325
Fa    35  
Gd    9   
Po    7   
Ex    2   
Name: Garage_Condition, dtype: int64

In [75]:
data.Garage_Quality.value_counts()

TA    1310
Fa    48  
Gd    14  
Ex    3   
Po    3   
Name: Garage_Quality, dtype: int64

In [76]:
data.Garage_Finish_Year.value_counts()

Unf    605
RFn    422
Fin    351
Name: Garage_Finish_Year, dtype: int64

In [77]:
data.Garage.value_counts()

Attchd     869
Detchd     387
BuiltIn    88 
Basment    19 
CarPort    9  
2TFes      5  
2Types     1  
Name: Garage, dtype: int64

## Imputation

In [78]:
from sklearn.impute import SimpleImputer

In [79]:
cat_imputer = SimpleImputer(strategy='most_frequent')

In [80]:
cat_imputer.fit(data[cat_null_cols])

data[cat_null_cols] = cat_imputer.transform(data[cat_null_cols])  

In [81]:
data['Lot_Extent'].mean()

70.04583333333333

In [82]:
data['Brick_Veneer_Area'].mean()

103.75671950379049

In [83]:
data['Garage_Built_Year'].mean()

1978.5159651669085

In [84]:
data['Lot_Extent'] = data['Lot_Extent'].replace(np.nan, 70)

data['Brick_Veneer_Area'] = data['Brick_Veneer_Area'].replace(np.nan, 100)

data['Garage_Built_Year'] = data['Garage_Built_Year'].replace(np.nan, 1978)

In [85]:
data.isnull().sum()

Building_Class               0
Zoning_Class                 0
Lot_Extent                   0
Lot_Size                     0
Road_Type                    0
Property_Shape               0
Land_Outline                 0
Utility_Type                 0
Lot_Configuration            0
Property_Slope               0
Neighborhood                 0
Condition1                   0
Condition2                   0
House_Type                   0
House_Design                 0
Overall_Material             0
House_Condition              0
Construction_Year            0
Remodel_Year                 0
Roof_Design                  0
Roof_Quality                 0
Exterior1st                  0
Exterior2nd                  0
Brick_Veneer_Type            0
Brick_Veneer_Area            0
Exterior_Material            0
Exterior_Condition           0
Foundation_Type              0
Basement_Height              0
Basement_Condition           0
Exposure_Level               0
BsmtFinType1                 0
BsmtFinS

In [86]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [87]:
data[data.select_dtypes(include = ['object']).columns] = data[data.select_dtypes(include = ['object']).columns].apply(le.fit_transform)

In [88]:
data.head()

Building_Class  Zoning_Class  Lot_Extent  Lot_Size  Road_Type  Property_Shape  Land_Outline  Utility_Type  Lot_Configuration  Property_Slope  Neighborhood  Condition1  Condition2  House_Type  House_Design  Overall_Material  House_Condition  Construction_Year  Remodel_Year  Roof_Design  Roof_Quality  Exterior1st  Exterior2nd  Brick_Veneer_Type  Brick_Veneer_Area  Exterior_Material  Exterior_Condition  Foundation_Type  Basement_Height  Basement_Condition  Exposure_Level  BsmtFinType1  BsmtFinSF1  BsmtFinType2  BsmtFinSF2  BsmtUnfSF  Total_Basement_Area  Heating_Type  Heating_Quality  Air_Conditioning  Electrical_System  First_Floor_Area  Second_Floor_Area  LowQualFinSF  Grade_Living_Area  Underground_Full_Bathroom  Underground_Half_Bathroom  Full_Bathroom_Above_Grade  Half_Bathroom_Above_Grade  Bedroom_Above_Grade  Kitchen_Above_Grade  Kitchen_Quality  Rooms_Above_Grade  Functional_Rate  Fireplaces  Garage  Garage_Built_Year  Garage_Finish_Year  Garage_Size  Garage_Area  Garage_Quality  Garage_Condition  Pavedd_Drive  W_Deck_Area  Open_Lobby_Area  Enclosed_Lobby_Area  Three_Season_Lobby_Area  Screen_Lobby_Area  Pool_Area  Miscellaneous_Value  Month_Sold  Year_Sold  Sale_Type  Sale_Condition  Sale_Price
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
1   60              3             65.0        8450      1          3               3             0             4                  0               5             2           2           0           5             7                 5                2003               2003          1            4             12           13           1                  196.0              2                  4                   2                2                3                   3               2             706         5             0           150        856                  1             0                1                 4                  856               854                0             1710               1                          0                          2                          1                          3                    1                    2                8                  7                0           2       2003.0             1                   2            1085.793744  4               4                 2             163.788080   69.596115        20.337934            0                        0                  0          0                    2           2008       8          4               208500    
2   20              3             80.0        9600      1          3               3             0             2                  0               24            1           2           0           2             6                 8                1976               1976          1            4             8            8            2

## Train-Validation Split

In [89]:
from sklearn.model_selection import train_test_split

train, valid = train_test_split(data, test_size = 0.2, random_state = 123)

In [90]:
print("Shape of training set:", train.shape)
print("Shape of validation set:", valid.shape)

Shape of training set: (1167, 75)
Shape of validation set: (292, 75)


In [91]:
x_train = train.drop('Sale_Price', axis = 1)
y_train = train['Sale_Price']

x_valid = valid.drop('Sale_Price', axis = 1)
y_valid = valid['Sale_Price']

In [92]:
print(x_train.shape)

print(x_valid.shape)

(1167, 74)
(292, 74)


## Model Building

### Linear Regression

In [93]:
from sklearn.linear_model import LinearRegression

LR = LinearRegression()

LR.fit(x_train, y_train)

pred_train = LR.predict(x_train)
pred_valid = LR.predict(x_valid)

Rsquare = LR.score(x_train, y_train)
Rsquare

0.848989257708592

In [94]:
from sklearn import metrics

In [95]:
print('Train RMSE:',np.sqrt(metrics.mean_squared_error(y_train, pred_train)))
print('Validation RMSE:',np.sqrt(metrics.mean_squared_error(y_valid, pred_valid)))

Train RMSE: 30937.554584136204
Validation RMSE: 28341.61254560808


In [96]:
metrics.mean_squared_error(y_train, pred_train)

957132283.646407

## Model Improvement

## Feature Scaling

In [97]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()

x_train_sc = sc.fit_transform(x_train)
x_valid_sc = sc.transform(x_valid)

In [98]:
x_train_sc[0]

array([ 0.06992307, -0.06352233,  0.95798691, -0.04646956,  0.07188852,
        0.7280666 , -1.07552059, -0.02928537,  0.60045217,  3.53115915,
        0.43727102, -0.06228172, -0.03550638, -0.41801589,  1.03358884,
        0.64974025, -0.51502886,  0.72912199,  0.39263475, -0.4997353 ,
       -0.05924372, -1.14349218, -1.23469376, -1.23240056,  0.85726811,
       -0.75184515,  0.3678275 ,  0.77595459, -0.31075024,  0.27649445,
       -1.16829396, -0.41367476,  1.15871945,  0.29624622, -0.28058847,
       -0.46725852,  0.63908912, -0.10549345, -0.87936835,  0.26393344,
        0.30553558,  0.51504639,  2.12438522, -0.12096907,  2.0967568 ,
        1.09941809, -0.23088841,  0.7837699 ,  1.24539532,  1.35856403,
       -0.22869055, -0.44631126,  1.49468591,  0.26717312,  0.61814608,
       -0.67162838,  0.61226776, -0.27999659,  1.66036254, -0.36756631,
        0.2225218 ,  0.18540391,  0.29260894, -0.02930076, -0.22604567,
       -0.41155572, -0.11797869, -0.25736283, -0.07687816, -0.08

In [99]:
LR = LinearRegression()

LR.fit(x_train_sc, y_train)

pred_train = LR.predict(x_train_sc)
pred_valid = LR.predict(x_valid_sc)

Rsquare = LR.score(x_train_sc, y_train)
Rsquare

0.8489808522641625

Here are three common evaluation metrics for regression problems:

**Mean Absolute Error** (MAE) is the mean of the absolute value of the errors:
 
**Mean Squared Error** (MSE) is the mean of the squared errors:
 
**Root Mean Squared Error** (RMSE) is the square root of the mean of the squared errors:

 
Comparing these metrics:

- **MAE** is the easiest to understand, because it's the average error.
- **MSE** is more popular than MAE, because MSE "punishes" larger errors, which tends to be useful in the real world.
- **RMSE** is even more popular than MSE, because RMSE is interpretable in the "y" units.

All of these are loss functions, because we want to minimize them.

We will be using **RMSE** loss function to evaluate our predictions.

In [100]:
print('Train RMSE:',np.sqrt(metrics.mean_squared_error(y_train, pred_train)))
print('Validation RMSE:',np.sqrt(metrics.mean_squared_error(y_valid, pred_valid))) 

Train RMSE: 30938.415583404865
Validation RMSE: 28339.92828840908


By using **Standard Scaler** method we were able to reduce **RMSE** by **50,000** units approx.

In [101]:
!pip install lazypredict

In [102]:
from lazypredict.Supervised import LazyRegressor

## Creating Regression Models

In [103]:
reg = LazyRegressor(verbose=0, ignore_warnings=False, custom_metric=None)

In [104]:
models, predictions = reg.fit(x_train_sc, x_valid_sc, y_train, y_valid)

 90%|█████████████████████████████████████████████████████████████████████████▌        | 35/39 [00:23<00:02,  1.35it/s]

StackingRegressor model failed to execute
__init__() missing 1 required positional argument: 'estimators'


100%|██████████████████████████████████████████████████████████████████████████████████| 39/39 [00:25<00:00,  1.55it/s]


In [105]:
models

R-Squared      RMSE  Time Taken
Model                                                         
GradientBoostingRegressor     0.91       24004.86  1.25       
LGBMRegressor                 0.89       25838.15  0.63       
XGBRegressor                  0.89       26567.55  1.44       
RandomForestRegressor         0.88       26848.07  3.50       
HistGradientBoostingRegressor 0.88       26911.95  4.71       
ExtraTreesRegressor           0.88       27429.61  2.90       
BaggingRegressor              0.87       28052.53  0.37       
HuberRegressor                0.87       28248.31  0.26       
LassoLars                     0.87       28317.14  0.07       
RidgeCV                       0.87       28340.27  0.08       
Lasso                         0.87       28341.01  0.17       
Ridge                         0.87       28341.34  0.03       
LinearRegression              0.87       28341.62  0.03       
TransformedTargetRegressor    0.87       28341.62  0.03       
BayesianRidge                 0.87       28373.16  0.25       
SGDRegressor                  0.87       28395.68  0.06       
Lars                          0.87       28630.98  0.10       
PassiveAggressiveRegressor    0.87       28656.77  0.35       
LassoCV                       0.86       28953.40  0.31       
LassoLarsCV                   0.86       29006.70  0.25       
ElasticNet                    0.86       29698.06  0.13       
LassoLarsIC                   0.85       30516.29  0.07       
OrthogonalMatchingPursuitCV   0.82       33221.33  0.05       
OrthogonalMatchingPursuit     0.82       33221.33  0.06       
AdaBoostRegressor             0.82       33439.29  0.92       
LarsCV                        0.81       34155.73  0.25       
RANSACRegressor               0.77       37812.03  0.39       
KNeighborsRegressor           0.75       39589.31  0.18       
DecisionTreeRegressor         0.71       42401.96  0.11       
ExtraTreeRegressor            0.70       43061.02  0.05       
ElasticNetCV                  0.18       71072.46  0.41       
DummyRegressor                -0.00      78741.76  0.02       
NuSVR                         -0.02      79465.11  0.30       
SVR                           -0.07      81453.80  0.43       
KernelRidge                   -4.36      182210.78 0.17       
MLPRegressor                  -5.22      196238.11 3.65       
LinearSVR                     -5.33      197999.10 0.03       
GaussianProcessRegressor      -5.38      198788.69 1.07

### Gradient Boosting Regression

In [106]:
from sklearn.ensemble import GradientBoostingRegressor

GB = GradientBoostingRegressor(random_state=0)

GB.fit(x_train_sc, y_train)

GB_pred_train = GB.predict(x_train_sc)
GB_pred_valid = GB.predict(x_valid_sc)

GB_Rsquared = GB.score(x_train_sc, y_train)
GB_Rsquared

0.9672460642179866

In [110]:
print('Gradient Boosting Regression', '\n')

print('Train RMSE:',np.sqrt(metrics.mean_squared_error(y_train, GB_pred_train)), '\n')

print('Validation RMSE:',np.sqrt(metrics.mean_squared_error(y_valid, GB_pred_valid))) 

Gradient Boosting Regression 

Train RMSE: 14408.335347957687 

Validation RMSE: 24240.914190109957


Our model is **Overfitting** as we can see **Low Training Error** representing **Low bias** and
**High validation Error** meaning **High Variance** due to the **Outliers** in our data.

In [111]:
GB.get_params(deep=True)

{'alpha': 0.9,
 'ccp_alpha': 0.0,
 'criterion': 'friedman_mse',
 'init': None,
 'learning_rate': 0.1,
 'loss': 'ls',
 'max_depth': 3,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_iter_no_change': None,
 'presort': 'deprecated',
 'random_state': 0,
 'subsample': 1.0,
 'tol': 0.0001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

## Cross-Validation

Cross-validation is primarily used in applied machine learning to estimate the skill of a machine learning model on unseen data.

### Hyperparameter Tuning Using Randomized Search CV

In [112]:
from sklearn.model_selection import RandomizedSearchCV

In [120]:
GB = GradientBoostingRegressor()

parameters = {'n_estimators':[50,75,100,125],
                   'criterion':['mse','mae'],
                   'max_depth':[2,3,5],
                   'min_samples_split':[2,3,4],
                   'min_samples_leaf':[1,2,3]}

Randomized_SearchCV_GB = RandomizedSearchCV(GB, parameters, cv = 5)

In [121]:
Randomized_SearchCV_GB.fit(x_train_sc, y_train)

RandomizedSearchCV(cv=5, error_score=nan,
                   estimator=GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0,
                                                       criterion='friedman_mse',
                                                       init=None,
                                                       learning_rate=0.1,
                                                       loss='ls', max_depth=3,
                                                       max_features=None,
                                                       max_leaf_nodes=None,
                                                       min_impurity_decrease=0.0,
                                                       min_impurity_split=None,
                                                       min_samples_leaf=1,
                                                       min_samples_split=2,
                                                       min_weight_fraction_leaf=0.0,
                             

In [122]:
Randomized_SearchCV_GB.best_params_

{'n_estimators': 125,
 'min_samples_split': 4,
 'min_samples_leaf': 1,
 'max_depth': 3,
 'criterion': 'mse'}

Applying above parameters

In [123]:
GB = GradientBoostingRegressor(n_estimators=125, criterion='mse', max_depth=3, min_samples_split=4, min_samples_leaf=1)

GB.fit(x_train_sc, y_train)

pred_train = GB.predict(x_train_sc)
pred_valid = GB.predict(x_valid_sc)


In [124]:
print("Gradient Boosting Regression after hyper-parameter tuning", '\n')

print('Train RMSE:',np.sqrt(metrics.mean_squared_error(y_train, pred_train)), '\n')

print('Validation RMSE:',np.sqrt(metrics.mean_squared_error(y_valid, pred_valid))) 

Gradient Boosting Regression after hyper-parameter tuning 

Train RMSE: 13570.697730167552 

Validation RMSE: 24070.35482607438


In [125]:
GB.score(x_train_sc, y_train)  # R-Squared

0.9709437052890209

### Hyperparameter Tuning Using Grid Search CV

In [126]:
from sklearn.model_selection import GridSearchCV

Grid_SearchCV_GB.fit(x_train_sc, y_train)

In [130]:
Grid_SearchCV_GB.best_params_

{'criterion': 'mae',
 'max_depth': 3,
 'min_samples_leaf': 1,
 'min_samples_split': 3,
 'n_estimators': 125}

Applying above parameters

In [131]:
GB = GradientBoostingRegressor(n_estimators=125, criterion='mae', max_depth=3, min_samples_split=3, min_samples_leaf=1)

GB.fit(x_train_sc, y_train)

pred_train = GB.predict(x_train_sc)
pred_valid = GB.predict(x_valid_sc)

In [132]:
print("Gradient Boosting Regression after hyper-parameter tuning using Grid Search CV", '\n')

print('Train RMSE:',np.sqrt(metrics.mean_squared_error(y_train, pred_train)), '\n')

print('Validation RMSE:',np.sqrt(metrics.mean_squared_error(y_valid, pred_valid))) 

Gradient Boosting Regression after hyper-parameter tuning using Grid Search CV 

Train RMSE: 14705.351809778635 

Validation RMSE: 23668.708627144573


In [133]:
GB.score(x_train_sc, y_train)  # R-Squared

0.9658817525022466

## Conclusion

**R-squared** remains **unchanged** at **0.97** even after hyper-parameter tuning using Random serach CV and Grid search CV.

**Gradient Boosting Regression Model** is still **overfitting** (Low Bias and High variance) due to the **Outliers**.

**Linear Regression** is the most **generalized model** as compare to Gradient Boosting model.